DOCSP-43056 Run a command (#64)
mongodb/docs-c: source/run-command.txt

   explicity pass it as an argument. The following code demonstrates how to
   specify a read preference and use it with the
   ``mongoc_client_command_simple()`` function:

   .. code-block:: c

      read_prefs = mongoc_read_prefs_new(MONGOC_READ_SECONDARY);

      command = BCON_NEW("hello", BCON_INT32(1));

      retval = mongoc_client_command_simple(client, "admin", command,
      read_prefs, NULL, &reply, &error);
  
   For more information on read preference options, see :manual:`Read Preference
   </core/read-preference/>` in the Server manual.

Response
--------

Each function returns either a BSON document or a cursor containing the response
from the database after the command runs. Each database command performs a
different function, so the response content can vary across commands. However,
every response contains documents with the following fields:

- ``<command result>``: Provides fields specific to the database command. For
  example, ``count`` returns the ``n`` field and ``explain`` returns the
  ``queryPlanner`` field.
- ``ok``: Indicates whether the command succeeded (``1``) or failed (``0``).

Example
-------

The following code shows how you can use the

> ``mongoc_client_command_with_opts()`` function to run the
> ``cloneCollectionAsCapped`` command with the ``writeConcern`` option, then run

Reviewer: S: Divide this into two since that sentence is pretty long
```suggestion
``mongoc_client_command_with_opts()`` function to run the
``cloneCollectionAsCapped`` command with the ``writeConcern`` option. It then runs
```

the ``distinct`` command with the ``collation`` and ``readConcern`` options.

.. literalinclude:: /includes/usage-examples/run-command.c
    :language: c
    :dedent:
    :emphasize-lines: 95-98, 100, 112, 128, 131-133, 135-138
    :start-after: start-run-command-example
    :end-before: end-run-command-example

Output
~~~~~~

The ``cloneCollectionAsCapped`` command clones a collection as a capped
collection. Then, the ``distinct`` command gets the distinct values of a field

> with a given filter and collation.

Reviewer: S: Introduce the code showing the output
```suggestion
with a given filter and collation. The example outputs the following result:
```

.. _c-run-command:

======================

> Run a Database Command

Author: [note to reviewer] Added "Database" to this title as this page will live in the `/databases-collections` drawer.

======================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :description: Learn how to use the C Driver to run a database command.
   :keywords: command, server, usage, code example

Overview
--------

In this guide, you can learn how to run a database command with the
{+driver-short+}. You can use database commands to perform a variety of
administrative and diagnostic tasks, such as fetching server statistics,
initializing a replica set, or running an aggregation pipeline.

.. important:: Prefer Driver Methods to Database Commands

   The driver provides wrapper methods for many database commands. We recommend

>    using driver methods instead of executing database commands when possible.

Reviewer: We refer to them as "functions" rather than "methods" in the C driver. Applies to all instances
```suggestion
   using driver functions instead of executing database commands when possible.
```

   
   To perform administrative tasks, use the :mongosh:`MongoDB Shell </>` instead
   of the {+driver-short+}. Calling the ``db.runCommand()`` method inside the
   shell is the preferred method to issue database commands, as it provides a
   consistent interface between the shell and drivers.

Execute a Command
-----------------

To run a database command, you must specify the command and any relevant
parameters in a BSON document, then pass this BSON document to a command
execution method. The {+driver-short+} provides the following methods to run
database commands:

- ``mongoc_client_command_simple()``, which returns the first document from the

>   result cursor into ``reply``. This method is a simplified interface that sends

Reviewer: I'm a little confused at what this means still. What is it returning documents from? and what is a ``reply``? Can this maybe be rephrased or expanded on a bit?

Author: Rewording!

  a raw command to the server.
- ``mongoc_client_command_with_opts()``, which runs a command on the server and

>   interprets ``opts`` according to this MongoDB server version. This method

Reviewer: ```suggestion
  interprets ``opts`` according to the MongoDB server version. This method
```

  server and interprets ``opts`` according to the MongoDB server version. This

>   function offers flexibiilty by allowing additional options.

Reviewer 2: ```suggestion
  function offers flexibility by allowing additional options.
```

The following code shows how you can use the ``mongoc_client_command_simple()``
method to run the ``hello`` command, which returns information about the current
member's role in the replica set, on a database:

.. literalinclude:: /includes/usage-examples/run-command.c
    :language: c
    :dedent:
    :start-after: start-execute-command
    :end-before: end-execute-command

For a full list of database commands and corresponding parameters, see the
:ref:`Additional Information <c-addl-info-runcommand>` section.

Command Options
---------------

You can specify optional command behavior for the
``mongoc_client_command_with_opts()`` method.

The ``mongoc_client_command_with_opts()`` method accepts a BSON document for the

> ``opts`` parameter. This can include command options such as read concerns,
> write concerns, and session options. 

Reviewer: S: I think this last sentence can be removed since you're listing them all out just below
```suggestion
``opts`` parameter.
```

Author: Makes sense, editing.

> You can pass a BSON document with the following options:

Reviewer: ```suggestion
You can pass a BSON document the specifies the following options:
```

- ``readConcern``
- ``writeConcern``
- ``sessionId``
- ``collation``
- ``serverId``

To learn more about a command and the options that it accepts, locate the
command and follow the link in the :manual:`Database Commands
</reference/command/>` section of the Server manual.

> The following code shows you how to specify a ``grantRolesToUsers`` command with a

Reviewer: ```suggestion
The following code shows how to specify a ``grantRolesToUsers`` command with a
```

.. literalinclude:: /includes/usage-examples/run-command.c
    :language: c
    :dedent:
    :start-after: start-execute-command-with-options
    :end-before: end-execute-command-with-options

.. note:: Read Preference

   The ``mongoc_client_command_simple()`` and
   ``mongoc_client_command_with_opts()`` functions ignore the read preference

>    setting you might have set on your database. By default, these functions use

Reviewer 2: ```suggestion
   setting you might have set on your client. By default, these functions use
```

   ``mongoc_client_command_with_opts()`` methods ignore the read preference

>    setting you may have set on your ``Db`` object. By default, these methods use

Reviewer: Style guide: Use "might" instead of "may" when indicating possibility: https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#std-term-may
```suggestion
   setting you might have set on your ``Db`` object. By default, these methods use
```

>    setting you may have set on your ``Db`` object. By default, these methods use

Reviewer: ```suggestion
   setting you may have set on your database. By default, these methods use
```

   the ``primary`` read preference.

>    To specify a read preference other than the primary read preference, you need
>    to explicity pass it as an argument. The following code demonstrates how to

Reviewer: Style Guide - use "must" instead of "need to": https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#std-term-must
```suggestion
   To specify a read preference other than the primary read preference, you must explicity pass it as an argument. The following code demonstrates how to
```

   specify a read preference and use it with the
   ``mongoc_client_command_simple()`` method:

>    .. code-block:: C

Author: [note to reviewer] The highlighting for this codeblock shows up in my IDE but don't see it computed in the preview. Any tips to fix this are welcome!

Reviewer: I'm not 100% sure but it could be a casing issue. I think snooty expects a lowercase "c"

      command = BCON_NEW("hello", BCON_INT32(1));

>       retval = mongoc_client_command_simple(client, "admin", command,
>       read_prefs, NULL, &reply, &error);

Reviewer 2: ```suggestion
      retval = mongoc_client_command_simple(client, "admin", command,
      read_prefs, &reply, &error);
```

      retval = mongoc_client_command_simple(client, "admin", command,
      read_prefs, NULL, &reply, &error);
  
   For more information on read preference options, see :manual:`Read Preference
   </core/read-preference/>` in the Server manual.

Response
--------

Each method returns either a BSON document or a cursor containing the response
from the database after the command runs. Each database command performs a
different function, so the response content can vary across commands. However,
every response contains documents with the following fields:

> - **<command result>:** Provides fields specific to the database command. For

Reviewer: S: I don't think this should be bolded, maybe monospaced instead?
```suggestion
- <command result>: Provides fields specific to the database command. For
```

- ``ok``: Indicates whether the command succeeded (``1``) or failed (``0``).

Example
-------

The following code shows how you can use the

> ``mongoc_client_command_with_opts()`` function to run the

Reviewer 2: Suggest revising. Example uses `mongoc_client_write_command_with_opts` and `mongoc_client_read_command_with_opts`, not `mongoc_client_command_with_opts`.

The following code shows how you can use the
``mongoc_client_command_with_opts()`` method to run the
``cloneCollectionAsCapped`` command with the ``writeConcern`` option, then run

> the ``distinct`` command with the ``collation`` and ``readConcern`` options.

Reviewer: Since this example is so long, can we highlight the relevant bits? At least the methods/options that are called out in this paragraph

Author: Makes sense! I added the lines to emphasize but not seeing the highlighting in the preview, lmk if you can tell what the include is missing 😅

Reviewer: It looks like you added the actual line numbers from the file, but I think the `:emphasize-lines:` is relative to the included code block, not the actual file. So the first line that gets included would be line 1, regardless of what line it is in the file

Author: Ah ok that makes sense! Thank you 🙏
